crates/polars-expr/src/expressions/eval.rs

```diff
@@ -276,6 +281,134 @@ impl EvalExpr {
         }
     }
 
+    #[cfg(feature = "dtype-array")]
+    fn evaluate_on_array_chunked(
+        &self,
+        ca: &ArrayChunked,
+        state: &ExecutionState,
+        as_list: bool,
+    ) -> PolarsResult<Column> {
+        let df = ca.get_inner().with_name(PlSmallStr::EMPTY).into_frame();
+
+        // Fast path: Empty or only nulls.
+        if ca.null_count() == ca.len() {
+            let name = self.output_field_with_ctx.name.clone();
+            let dtype = self.non_aggregated_output_dtype.inner_dtype().unwrap();
+
+            return Ok(if as_list {
+                ListChunked::full_null_with_dtype(name, ca.len(), dtype).into_column()
+            } else {
+                ArrayChunked::full_null_with_dtype(name, ca.len(), dtype, ca.width()).into_column()
+            });
+        }
+
+        // Fast path: fully elementwise expression without masked out values.
+        if self.evaluation_is_elementwise && !ca.has_nulls() {
+            let column = self.evaluation.evaluate(&df, state)?;
+            assert_eq!(column.len(), ca.len() * ca.width());
+
+            let dtype = column.dtype().clone();
+            let out = ArrayChunked::from_aligned_values(
+                self.output_field_with_ctx.name.clone(),
+                &dtype,
+                ca.width(),
+                column.take_materialized_series().into_chunks(),
+                ca.len(),
+            );
+
+            return Ok(if as_list {
+                out.to_list().into_column()
+            } else {
+                out.clone().into_column()
+            });
+        }
+
+        let validity = ca.rechunk_validity();
+
+        // Create groups for all valid array elements.
+        let groups = if ca.has_nulls() {
+            let validity = validity.as_ref().unwrap();
+            (0..ca.len())
+                .filter(|i| unsafe { validity.get_bit_unchecked(*i) })
+                .map(|i| [(i * ca.width()) as IdxSize, ca.width() as IdxSize])
+                .collect()
+        } else {
+            (0..ca.len())
+                .map(|i| [(i * ca.width()) as IdxSize, ca.width() as IdxSize])
```

Comment:
Where do we check if `len * width < IdxSize::MAX`?

Reply:
I discussed this with @orlp and I don't think it is really possible to provide such a guarantee without massively sacrificing performance for the normal case. On input data you could check this yes, but you are not able to predict what the output size will be. Consequently, you can overflow the `IdxSize` at any given point.

@orlp and I agreed that it is probably best to adopt the policy of limiting the `IdxSize` to the *leaf* values and not to the *root* values. We don't do this properly basically anywhere and requires so much additional work and cost to base performance that I don't think it is worth it.